lib/spack/spack/build_environment.py

```diff
@@ -1295,14 +1296,17 @@ def exitcode_msg(p):
         typ = "exit" if p.exitcode >= 0 else "signal"
         return f"{typ} {abs(p.exitcode)}"
 
+    p.join(timeout=timeout)
+    if p.is_alive():
+        warnings.warn(f"Terminating process, since the timeout of {timeout}s was exceeded")
+        p.terminate()
+        p.join()  # Ensure the process has fully stopped
```

Comment:
See here for how we did this in the concurrent build PR: https://github.com/spack/spack/pull/47590/files#diff-a493ca2ca45c40f1a67e6ef60fe57654cb2077e3d20b50ceee9b85d0ec16c8e3R1113-R1123

you need to ask nicely with terminate (SIGTERM) and then kill (SIGKILL) if nothing happens.

Reply:
I added this logic, and also added a unit test in https://github.com/spack/spack/pull/49839/commits/d077838bc486cfc55ca3f4a407280e02357c0c22 I didn't go to a full refactor, and injected the process class - but if people like the idea and is not interfering too much with #47590 I can also do that If instead d077838bc486cfc55ca3f4a407280e02357c0c22 seems too much code for a unit test, I can revert it.

Reply:
I think we can work around this when rebasing 47590... it's nearly done and I think we can get it merged next week.